samples/msal-node-samples/EndToEndTestApp/app/test/device-code-aad/device-code-aad.spec.ts

```diff
@@ -0,0 +1,143 @@
+import "jest";
+import puppeteer from "puppeteer";
+import { ChildProcessWithoutNullStreams, spawn } from "child_process";
+import { Screenshot, createFolder, setupCredentials } from "../../../../../e2eTestUtils/TestUtils";
+import { NodeCacheTestUtils } from "../../../../../e2eTestUtils/NodeCacheTestUtils";
+import { LabClient } from "../../../../../e2eTestUtils/LabClient";
+import { LabApiQueryParams } from "../../../../../e2eTestUtils/LabApiQueryParams";
+import { AppTypes, AzureEnvironments } from "../../../../../e2eTestUtils/Constants";
+import { 
+    enterCredentials, 
+    enterDeviceCode,
+    SCREENSHOT_BASE_FOLDER_NAME,
+    extractDeviceCodeParameters,
+    validateCacheLocation,
+    checkTimeoutError
+ } from "../testUtils";
+
+const TEST_CACHE_LOCATION = `${__dirname}/data/testCache.json`;
+
+let username: string;
+let accountPwd: string;
+
+describe('Device Code AAD PPE Tests', () => {
+    jest.setTimeout(90000);
+
+    let browser: puppeteer.Browser;
+    let context: puppeteer.BrowserContext;
+    let page: puppeteer.Page;
+    let device: ChildProcessWithoutNullStreams;
+    const stream: Array<any> = [];
+    
+    beforeAll(async () => {
+        await validateCacheLocation(TEST_CACHE_LOCATION);
+        createFolder(SCREENSHOT_BASE_FOLDER_NAME);
+        const labApiParms: LabApiQueryParams = {
+            azureEnvironment: AzureEnvironments.PPE,
+            appType: AppTypes.CLOUD,
+        };
+
+        const labClient = new LabClient();
+        const envResponse = await labClient.getVarsByCloudEnvironment(labApiParms);
+        [username, accountPwd] = await setupCredentials(envResponse[0], labClient);
+
+        browser = await puppeteer.launch({
+            headless: true,
+            ignoreDefaultArgs: ['--no-sandbox', '-disable-setuid-sandbox', '--disable-extensions']
+        });
+    });
+
+    afterAll(async () => {
+        await browser.close();
+    });
+
+    describe("Acquire Token", () => {
+        let testName: string;
+        let screenshot: Screenshot;
+
+        beforeAll(async () => {
+            await NodeCacheTestUtils.resetCache(TEST_CACHE_LOCATION);
+        });
+
+        beforeEach(async () => {
+            device = spawn(/^win/.test(process.platform) ? "npm.cmd" : "npm", ["start", "--", "-s", "device-code-aad", "-c", TEST_CACHE_LOCATION]);
```

Comment:
Do you have to spin up a new server for each test? May be faster to spin up a single server per test suite (and this is something my recent work for react does as part of the setup. It may be worthwhile to implement those changes here as well)

Reply:
We can't reuse a server for multiple tests as once a server gets the required user authentication or a cancellation the sever terminates, thus reducing our ability to reuse or reboot the server instance.

Reply:
Per @tnorling's advice, I replaced the child processes by pulling the route logic directly into the test file and triggering the device code token acquisition from within the tests.